aws/request.go

```diff
@@ -217,3 +227,99 @@ func (r *Request) Send() error {
 
 	return nil
 }
+
+func (r *Request) HasNextPage() bool {
+	return r.nextPageTokens() != nil
+}
+
+func (r *Request) nextPageTokens() []interface{} {
+	if r.Operation.Paginator == nil {
+		return nil
+	}
+
+	if r.Operation.TruncationToken != "" {
+		tr := awsutil.ValuesAtAnyPath(r.Data, r.Operation.TruncationToken)
+		if tr == nil || len(tr) == 0 {
+			return nil
+		}
+		switch v := tr[0].(type) {
+		case bool:
+			if v == false {
+				return nil
+			}
+		}
+	}
+
+	found := false
+	tokens := make([]interface{}, len(r.Operation.OutputTokens))
+
+	for i, outtok := range r.Operation.OutputTokens {
+		v := awsutil.ValuesAtAnyPath(r.Data, outtok)
+		if v != nil && len(v) > 0 {
+			found = true
+			tokens[i] = v[0]
+		}
+	}
+
+	if found {
+		return tokens
+	}
+	return nil
+}
+
+func (r *Request) NextPage() *Request {
+	tokens := r.nextPageTokens()
+	if tokens == nil {
+		return nil
+	}
+
+	data := reflect.New(reflect.TypeOf(r.Data).Elem()).Interface()
+	nr := NewRequest(r.Service, r.Operation, awsutil.CopyOf(r.Params), data)
+	for i, intok := range nr.Operation.InputTokens {
+		awsutil.SetValueAtAnyPath(nr.Params, intok, tokens[i])
+	}
+	return nr
+}
+
+// This nil value is needed to work around passing nil into reflect.Value#Call()
+// See https://groups.google.com/forum/#!topic/golang-nuts/apNcACpl_fI
+var nilval interface{}
+var rnil = reflect.ValueOf(&nilval).Elem()
+
+func (r *Request) EachPage(fn interface{}) error {
+	valfn := reflect.ValueOf(fn)
+	if valfn.Kind() != reflect.Func {
+		panic("expected function for EachPage()")
+	}
+	if valfn.Type().NumOut() > 0 && valfn.Type().Out(0).Kind() != reflect.Bool {
```

Comment:
I would check for "exactly one output parameter" (`valfn.Type().NumOut() == 1`) to enforce stronger typing.

While think about it, the valid type for this function is actually
```go
func (data SomeType, more bool) bool
```
Only `SomeType` cannot be represented properly by Go at the moment.

An alternate design might be to use 
```go
func (data interface{}, more bool) bool
```

That would make this function much safer to use, but maybe a little less convenient. 
But it will also get rid of reflection here.

It might be possible to hide this better by providing type asserting helpers for the specific use cases and/or more code generation.